File: lib/biotp-py/biotp/android_attest.py
"""
Android Key Attestation verification.

Verifies that an Android KeyStore attestation certificate chain is rooted
to a Google hardware attestation root and that the key attestation extension
(OID 1.3.6.1.4.1.11129.2.1.17) contains the expected challenge.
"""

import base64
import hashlib
import hmac
import time
from typing import Optional

from cryptography import x509
from cryptography.hazmat.primitives.asymmetric import ec, rsa, padding
from cryptography.hazmat.primitives import hashes
from cryptography.x509.oid import ObjectIdentifier

# Google hardware attestation root certificates (there are two known roots).
# We accept either.
GOOGLE_ROOT_CERTS_PEM = [
    # Root 1 (RSA)
    b"""\
-----BEGIN CERTIFICATE-----
MIIFYDCCBEigAwIBAgIJAOj6GWMU0voYMA0GCSqGSIb3DQEBCwUAMBsxGTAXBgNV
BAUTEGY0NzY2OTMzMjgwNDQxNDEwHhcNMTIwNDI1MTE0MzI3WhcNMjIwNDIzMTE0
MzI3WjAbMRkwFwYDVQQFExBmOTIwMDllODUzYjZiMDQ1MIICIjANBgkqhkiG9w0B
AQEFAAOCAg8AMIICCgKCAgEAr7bHgiuxpwHsK7Qui8xUFmOr75gvMsd/dTEDDJdS
Sxtf6An7xyqpRR90PL2abxM1dEqlXnf2tqw1Ne4Xwl5jlRfdnJLmN0pTy/4lj4/7
tv0Sk3iiKkypnEUtR6WfMgH0QZfKHM1+di+y9TFRtv6y//0rb+T+W8a9nsNL/ggj
nar86461qO0rOs2cXjp3kOG1FEJ5MVmFmBGtnrKpa73XpXyTqRxB/M0n1n/W9nGq
C4FSYa04T6N5RIZGBN2z2MT5IKGbFlbC8UrW0DxW7AYImQQcHtGl/m00QLVWutHQ
oVJYnFPlXTcHYvASLu+RhhsbDmxMgJJ0mcDpvsC4PjvB+TxywElgS70vE0XmLD+O
JtvsBslHZvPBKCOdT0MS+tgSOIfga+z1Z1g7+DVagf7quvmag8jfPioyKvxnK/EgS
TyF15Iq/e7hjAolP1kmjXpgXrVXM3/QMXR0Cf5tGJwUbKZ3ruA2MkpJKfPRi9Y4b
gKz4z/qzXAJEppJ1S+IZAmNuSHRUgJiw0K/gI9BjfDyDr3P0b2RYUUU4JJslxUHh
E2skEz0DlCEQgGHCIxFgVKgvFSWwPUHUbUg9K/wDKrQDIxdfrVIaenUvMSbG0SRN
5UxDGJJH5mT/bNzYOSg/x7YRNhQ0jlApNJ3U4Zhl0RUM8QT8JHPO/9L8MIyR9/0
CggAUAADAQABo4IBOjCCATYwHQYDVR0OBBYEFMit6XdMRcOjzw0WEOR5QzohWjDPM
B8GA1UdIwQYMBaAFMit6XdMRcOjzw0WEOR5QzohWjDPMA8GA1UdEwEB/wQFMAMBAf
8wDgYDVR0PAQH/BAQDAgGGMDIGCCsGAQUFBwEBBCYwJDAiBggrBgEFBQcwAoYWaHR
0cDovL28ucGtpLmdvb2cvZ3NyMjAyBgNVHR8EKzApMCegJaAjhiFodHRwOi8vby5w
a2kuZ29vZy9nc3IyL2dzcjIuY3JsMA0GCSqGSIb3DQEBCwUAA4ICAQBbSRnhSzNF
a/Y4lIwwSdb1dFqGOrxS2MXOgpRBnYezHCjG4GY07gMRDAFKsSFlq8u9o/t6QZMR
9+bOY33BoV5dnGey7c0F/YZ6O72EPmD8A/k+8l4Oyn2xwXAS11KEsr0z/Yf46SqN
tynGRZfWRCJRjE/GniLi9vx4g4w3GasJPRhsb4a/8aqExshEZG+1tBCjSCFKJFNF
qaBTzay1f8TBbQUG8lkN8cf/PjfAXmiJBpRz0bJHZBdDXzOPC4Io4lds/P5v7qEU
NbG/RvB7WUVAP31N/UJoFeZuf2kXOHX7d3FKF3nS1qkpGiQMHa1v47VPl2P3RSKU
RkKHa/YFYJpnxbZ0OGTM5PsZnZKl8AR3BclJvi3M1+mPNRWSXQN7167J8bPNR1YQ
IqjAoSiZHaKIajqp/rBmIqQLxKLhMJ4pKl9E2dJet1gN0VbJiOJiRUouRMDSIKNw
JpajxQJ1hFMN70mFnw5GhVi4gkRJ5cJSTbZnz6N1OVGzCKeOk0BO5Z8SfPshvVfc
MN3qN+GXRZRT+lHGH8MRsGEmVbITmGR2vp2GFLBPSAl2vWxn0jKNR3S6MgVk9DO0
cJHi1fA/IvNBjFqGVCFCCSsJnJBoEvdIDwDOh9DGBGyCkLnmJAE/EhteIJsBnVGm
Eah2JhsEbLfBGJpSiaKCh8kH/EU2Ug==
-----END CERTIFICATE-----""",
    # Root 2 (EC, more common on modern devices)
    b"""\
-----BEGIN CERTIFICATE-----
MIICMDCCAbagAwIBAgIKFMGhLgAAAAACDTAKBggqhkjOPQQDAjAbMRkwFwYDVQQF
ExBmOTIwMDllODUzYjZiMDQ1MB4XDTE5MTEyMjIwMzc1OFoXDTM0MTEyMjIwMzc1
OFowOzE5MDcGA1UEBRMwNjM2MDc2ODRiODNiMTkzNjM3NTA1MjUyNzI5OTAzNjBl
YWYxOTNkNTdlODQ4MDUwWTATBgcqhkjOPQIBBggqhkjOPQMBBwNCAASeJFhLlGbP
QMIBkMGkse/aJMQ2PmRrtNkfnOMqW6s+GSqJJCmhkMVLo3IjG80arlPkQ5+sse1Q
DLjyQ/sMtlTmo4G6MIG3MB0GA1UdDgQWBBQ2YeEAfIgFKBRih5aXMvixtPewBjAfB
gNVHSMEGDAWgBTIrel3TEXDo88NFhDkeUM6IVowzzAPBgNVHRMBAf8EBTADAQH/MA
4GA1UdDwEB/wQEAwICBDBUBgNVHR8ETTBLMEmgR6BFhkNodHRwczovL2FuZHJvaWQ
uZ29vZ2xlYXBpcy5jb20vYXR0ZXN0YXRpb24vY3JsLzA2MzYwNzY4NGI4M2IxOTM2
Mzc1MDUwKBgKKwYBBAHWeQIBFQQaMBgCARkKAQACAQMKAQEECGFiY2RlZmdoMAoG
CCqGSM49BAMCA0gAMEUCIDadx4jX0NQtw77N8jnPQOqiTnkCONBQwN6Y1ryP9n0P
AiEAl5mB+GCvKN7EqtLfqBMEkuaFd4053AEi7+QZDCEvm7c=
-----END CERTIFICATE-----""",
]

ANDROID_KEY_ATTESTATION_OID = ObjectIdentifier("1.3.6.1.4.1.11129.2.1.17")


def _load_google_roots():
    roots = []
    for pem in GOOGLE_ROOT_CERTS_PEM:
        try:
            roots.append(x509.load_pem_x509_certificate(pem))
        except Exception:
            pass
    return roots


GOOGLE_ROOTS = _load_google_roots()


def _safe_cert_name(name) -> str:
    try:
        if name is None:
            return "n/a"
        return ", ".join(f"{attr.oid._name}={attr.value}" for attr in name)
    except Exception:
        try:
            return name.rfc4514_string()
        except Exception:
            return "unable-to-parse"


def _verify_cert_signature(child, issuer_pub) -> bool:
    """Return True if child cert is signed by issuer_pub."""
    try:
        if isinstance(issuer_pub, ec.EllipticCurvePublicKey):
            issuer_pub.verify(
                child.signature,
                child.tbs_certificate_bytes,
                ec.ECDSA(child.signature_hash_algorithm),
            )
        elif isinstance(issuer_pub, rsa.RSAPublicKey):
            issuer_pub.verify(
                child.signature,
                child.tbs_certificate_bytes,
                padding.PKCS1v15(),
                child.signature_hash_algorithm,
            )
        else:
            return False
        return True
    except Exception:
        return False


def serialize_cert_chain(certs: list, root_label: str = "Google Root") -> list[dict]:
    """Serialize a certificate chain with per-link verification status."""
    if not certs:
        return []

    result = []
    for i, cert in enumerate(certs):
        try:
            subject = _safe_cert_name(cert.subject)
            issuer = _safe_cert_name(cert.issuer)
            serial = str(cert.serial_number)
            not_before = cert.not_valid_before_utc.isoformat() if hasattr(cert, 'not_valid_before_utc') else cert.not_valid_before.isoformat()
            not_after = cert.not_valid_after_utc.isoformat() if hasattr(cert, 'not_valid_after_utc') else cert.not_valid_after.isoformat()
        except Exception:
            subject = issuer = serial = not_before = not_after = "parse error"

        # Determine who signed this cert and whether it verifies
        if i < len(certs) - 1:
            signed_by = f"cert[{i+1}]"
            verified = _verify_cert_signature(cert, certs[i + 1].public_key())
        else:
            signed_by = root_label
            verified = False  # caller should set this for the root check

        entry = {
            "index": i,
            "role": "leaf" if i == 0 else ("root" if i == len(certs) - 1 else "intermediate"),
            "subject": subject,
            "issuer": issuer,
            "serial": serial,
            "not_before": not_before,
            "not_after": not_after,
            "signed_by": signed_by,
            "signature_valid": verified,
        }
        result.append(entry)

    return result


def _verify_chain(certs: list) -> tuple[bool, str, list[dict]]:
    """Verify each cert is signed by the next, and last is signed by a Google root.

    Returns (ok, reason, chain_info).
    """
    chain_info = serialize_cert_chain(certs, root_label="Google Hardware Attestation Root")
    if not certs:
        return False, "empty certificate chain", chain_info

    # Check intermediate links (already computed in serialize_cert_chain)
    for entry in chain_info[:-1]:
        if not entry["signature_valid"]:
            return False, f"chain verification failed at cert[{entry['index']}]", chain_info

    # Verify last cert against any known Google root (cryptographic check only)
    last_cert = certs[-1]
    root_verified = False
    for root in GOOGLE_ROOTS:
        if _verify_cert_signature(last_cert, root.public_key()):
            root_verified = True
            break

    if chain_info:
        chain_info[-1]["signature_valid"] = root_verified

    if not root_verified:
        return False, "chain not rooted to a known Google hardware attestation root", chain_info

    return True, "", chain_info


def verify_android_attestation(
    cert_chain_b64: list[str],
    expected_challenge: bytes,
    user_public_key_hex: str,
) -> tuple[bool, str, dict]:
    """Verify an Android KeyStore attestation certificate chain.

    Returns (ok, reason, details).
    """
    details: dict = {}

    if not cert_chain_b64:
        return False, "no attestation chain provided", details

    # Parse certs
    try:
        certs = []
        for i, b64 in enumerate(cert_chain_b64):
            der = base64.b64decode(b64)
            certs.append(x509.load_der_x509_certificate(der))
        details["chain_length"] = len(certs)
    except Exception as exc:
        return False, f"invalid certificate in chain: {exc}", details

    # Verify chain
    chain_ok, chain_reason, chain_info = _verify_chain(certs)
    details["chain_verified"] = chain_ok
    details["cert_chain"] = chain_info
    if not chain_ok:
        details["failure_reason"] = chain_reason
        return False, chain_reason, details

    # Extract key attestation extension from leaf
    leaf = certs[0]
    try:
        ext = leaf.extensions.get_extension_for_oid(ANDROID_KEY_ATTESTATION_OID)
        ext_data = ext.value.value if hasattr(ext.value, "value") else bytes(ext.value)
        details["attestation_extension_present"] = True
        details["attestation_extension_len"] = len(ext_data)
    except x509.ExtensionNotFound:
        details["attestation_extension_present"] = False
        return False, "no key attestation extension; key is not hardware-attested", details

    # Verify the leaf certificate public key matches the registered public key.
    # The leaf cert in an Android KeyStore attestation chain contains the attested key.
    try:
        leaf_pub = leaf.public_key()
        if isinstance(leaf_pub, ec.EllipticCurvePublicKey):
            leaf_nums = leaf_pub.public_numbers()
            leaf_x = leaf_nums.x.to_bytes(32, "big")
            leaf_y = leaf_nums.y.to_bytes(32, "big")
            leaf_raw_hex = (leaf_x + leaf_y).hex()
            details["leaf_public_key_hex"] = leaf_raw_hex[:20] + "..."
            if leaf_raw_hex != user_public_key_hex:
                return False, "leaf certificate public key does not match registered key", details
            details["public_key_binding"] = True
        else:
            return False, "leaf certificate does not contain an EC public key", details
    except Exception as exc:
        return False, f"failed to extract leaf public key: {exc}", details

    # Verify the challenge is present in the attestation extension.
    challenge_hash = hashlib.sha256(
        expected_challenge + bytes.fromhex(user_public_key_hex)
    ).digest()

    if expected_challenge in ext_data:
        details["challenge_binding"] = "raw_challenge_found"
    elif challenge_hash in ext_data:
        details["challenge_binding"] = "challenge_hash_found"
    else:
        return False, "challenge not found in attestation extension", details

    details["valid"] = True
    details["verified_at"] = time.time()
    return True, "ok", details

File: lib/biotp-ts/src/android_attest.ts
import * as crypto from "node:crypto";
import {
  findExtensionValue,
  verifyCertSignature,
  serializeCertChain,
  CertChainEntry,
} from "./x509";

const ANDROID_KEY_ATTESTATION_OID = "1.3.6.1.4.1.11129.2.1.17";

const TAG_USER_AUTH_TYPE = 504;
const TAG_NO_AUTH_REQUIRED = 503;
const TAG_AUTH_TIMEOUT = 505;
const AUTH_TYPE_FINGERPRINT = 0x02;
const SECURITY_LEVEL_TEE = 1;
const SECURITY_LEVEL_STRONGBOX = 2;

// -- DER parser for KeyDescription -------------------------------------

function parseDerElement(
  data: Buffer,
  offset: number
): { tag: number; constructed: boolean; cls: number; value: Buffer; end: number } {
  const tagByte = data[offset++];
  const cls = (tagByte >> 6) & 0x03;
  const constructed = (tagByte & 0x20) !== 0;
  let tag = tagByte & 0x1f;

  if (tag === 0x1f) {
    tag = 0;
    while (true) {
      const b = data[offset++];
      tag = (tag << 7) | (b & 0x7f);
      if (!(b & 0x80)) break;
    }
  }

  const lenByte = data[offset++];
  let length: number;
  if (lenByte < 0x80) {
    length = lenByte;
  } else {
    const n = lenByte & 0x7f;
    length = 0;
    for (let i = 0; i < n; i++) length = length * 256 + data[offset++];
  }

  const value = data.subarray(offset, offset + length);
  return { tag, constructed, cls, value, end: offset + length };
}

function parseDerSequence(
  data: Buffer
): Array<{ tag: number; constructed: boolean; cls: number; value: Buffer }> {
  const elements: Array<{ tag: number; constructed: boolean; cls: number; value: Buffer }> = [];
  let offset = 0;
  while (offset < data.length) {
    const el = parseDerElement(data, offset);
    elements.push({ tag: el.tag, constructed: el.constructed, cls: el.cls, value: el.value });
    offset = el.end;
  }
  return elements;
}

function parseDerInteger(data: Buffer): number {
  let val = data[0] & 0x80 ? -1 : 0;
  for (const b of data) val = (val << 8) | b;
  return val;
}

function parseAuthorizationList(seqData: Buffer): Record<number, unknown> {
  const result: Record<number, unknown> = {};
  for (const el of parseDerSequence(seqData)) {
    if (el.cls === 2 && el.constructed && el.value.length > 0) {
      const inner = parseDerElement(el.value, 0);
      if (inner.tag === 2) {
        result[el.tag] = parseDerInteger(inner.value);
      } else if (inner.tag === 5) {
        result[el.tag] = true;
      } else if (inner.tag === 17) {
        // SET OF INTEGER
        const items = parseDerSequence(inner.value);
        result[el.tag] = items.filter((i) => i.tag === 2).map((i) => parseDerInteger(i.value));
      } else {
        result[el.tag] = inner.value;
      }
    }
  }
  return result;
}

interface KeyDescription {
  attestationSecurityLevel: number;
  softwareEnforced: Record<number, unknown>;
  teeEnforced: Record<number, unknown>;
  error?: string;
}

function parseKeyDescription(extData: Buffer): KeyDescription {
  try {
    const outer = parseDerElement(extData, 0);
    if (outer.tag !== 16 || !outer.constructed)
      return { attestationSecurityLevel: 0, softwareEnforced: {}, teeEnforced: {}, error: "not a SEQUENCE" };

    const elements = parseDerSequence(outer.value);
    if (elements.length < 8)
      return { attestationSecurityLevel: 0, softwareEnforced: {}, teeEnforced: {}, error: `expected 8 elements, got ${elements.length}` };

    return {
      attestationSecurityLevel: parseDerInteger(elements[1].value),
      softwareEnforced: parseAuthorizationList(elements[6].value),
      teeEnforced: parseAuthorizationList(elements[7].value),
    };
  } catch (exc) {
    return { attestationSecurityLevel: 0, softwareEnforced: {}, teeEnforced: {}, error: String(exc) };
  }
}

const GOOGLE_ROOT_CERTS_PEM = [
  // Root 1 (RSA)
  `-----BEGIN CERTIFICATE-----
MIIFYDCCBEigAwIBAgIJAOj6GWMU0voYMA0GCSqGSIb3DQEBCwUAMBsxGTAXBgNV
BAUTEGY0NzY2OTMzMjgwNDQxNDEwHhcNMTIwNDI1MTE0MzI3WhcNMjIwNDIzMTE0
MzI3WjAbMRkwFwYDVQQFExBmOTIwMDllODUzYjZiMDQ1MIICIjANBgkqhkiG9w0B
AQEFAAOCAg8AMIICCgKCAgEAr7bHgiuxpwHsK7Qui8xUFmOr75gvMsd/dTEDDJdS
Sxtf6An7xyqpRR90PL2abxM1dEqlXnf2tqw1Ne4Xwl5jlRfdnJLmN0pTy/4lj4/7
tv0Sk3iiKkypnEUtR6WfMgH0QZfKHM1+di+y9TFRtv6y//0rb+T+W8a9nsNL/ggj
nar86461qO0rOs2cXjp3kOG1FEJ5MVmFmBGtnrKpa73XpXyTqRxB/M0n1n/W9nGq
C4FSYa04T6N5RIZGBN2z2MT5IKGbFlbC8UrW0DxW7AYImQQcHtGl/m00QLVWutHQ
oVJYnFPlXTcHYvASLu+RhhsbDmxMgJJ0mcDpvsC4PjvB+TxywElgS70vE0XmLD+O
JtvsBslHZvPBKCOdT0MS+tgSOIfga+z1Z1g7+DVagf7quvmag8jfPioyKvxnK/EgS
TyF15Iq/e7hjAolP1kmjXpgXrVXM3/QMXR0Cf5tGJwUbKZ3ruA2MkpJKfPRi9Y4b
gKz4z/qzXAJEppJ1S+IZAmNuSHRUgJiw0K/gI9BjfDyDr3P0b2RYUUU4JJslxUHh
E2skEz0DlCEQgGHCIxFgVKgvFSWwPUHUbUg9K/wDKrQDIxdfrVIaenUvMSbG0SRN
5UxDGJJH5mT/bNzYOSg/x7YRNhQ0jlApNJ3U4Zhl0RUM8QT8JHPO/9L8MIyR9/0
CggAUAADAQABo4IBOjCCATYwHQYDVR0OBBYEFMit6XdMRcOjzw0WEOR5QzohWjDPM
B8GA1UdIwQYMBaAFMit6XdMRcOjzw0WEOR5QzohWjDPMA8GA1UdEwEB/wQFMAMBAf
8wDgYDVR0PAQH/BAQDAgGGMDIGCCsGAQUFBwEBBCYwJDAiBggrBgEFBQcwAoYWaHR
0cDovL28ucGtpLmdvb2cvZ3NyMjAyBgNVHR8EKzApMCegJaAjhiFodHRwOi8vby5w
a2kuZ29vZy9nc3IyL2dzcjIuY3JsMA0GCSqGSIb3DQEBCwUAA4ICAQBbSRnhSzNF
a/Y4lIwwSdb1dFqGOrxS2MXOgpRBnYezHCjG4GY07gMRDAFKsSFlq8u9o/t6QZMR
9+bOY33BoV5dnGey7c0F/YZ6O72EPmD8A/k+8l4Oyn2xwXAS11KEsr0z/Yf46SqN
tynGRZfWRCJRjE/GniLi9vx4g4w3GasJPRhsb4a/8aqExshEZG+1tBCjSCFKJFNF
qaBTzay1f8TBbQUG8lkN8cf/PjfAXmiJBpRz0bJHZBdDXzOPC4Io4lds/P5v7qEU
NbG/RvB7WUVAP31N/UJoFeZuf2kXOHX7d3FKF3nS1qkpGiQMHa1v47VPl2P3RSKU
RkKHa/YFYJpnxbZ0OGTM5PsZnZKl8AR3BclJvi3M1+mPNRWSXQN7167J8bPNR1YQ
IqjAoSiZHaKIajqp/rBmIqQLxKLhMJ4pKl9E2dJet1gN0VbJiOJiRUouRMDSIKNw
JpajxQJ1hFMN70mFnw5GhVi4gkRJ5cJSTbZnz6N1OVGzCKeOk0BO5Z8SfPshvVfc
MN3qN+GXRZRT+lHGH8MRsGEmVbITmGR2vp2GFLBPSAl2vWxn0jKNR3S6MgVk9DO0
cJHi1fA/IvNBjFqGVCFCCSsJnJBoEvdIDwDOh9DGBGyCkLnmJAE/EhteIJsBnVGm
Eah2JhsEbLfBGJpSiaKCh8kH/EU2Ug==
-----END CERTIFICATE-----`,
  // Root 2 (EC, more common on modern devices)
  `-----BEGIN CERTIFICATE-----
MIICMDCCAbagAwIBAgIKFMGhLgAAAAACDTAKBggqhkjOPQQDAjAbMRkwFwYDVQQF
ExBmOTIwMDllODUzYjZiMDQ1MB4XDTE5MTEyMjIwMzc1OFoXDTM0MTEyMjIwMzc1
OFowOzE5MDcGA1UEBRMwNjM2MDc2ODRiODNiMTkzNjM3NTA1MjUyNzI5OTAzNjBl
YWYxOTNkNTdlODQ4MDUwWTATBgcqhkjOPQIBBggqhkjOPQMBBwNCAASeJFhLlGbP
QMIBkMGkse/aJMQ2PmRrtNkfnOMqW6s+GSqJJCmhkMVLo3IjG80arlPkQ5+sse1Q
DLjyQ/sMtlTmo4G6MIG3MB0GA1UdDgQWBBQ2YeEAfIgFKBRih5aXMvixtPewBjAfB
gNVHSMEGDAWgBTIrel3TEXDo88NFhDkeUM6IVowzzAPBgNVHRMBAf8EBTADAQH/MA
4GA1UdDwEB/wQEAwICBDBUBgNVHR8ETTBLMEmgR6BFhkNodHRwczovL2FuZHJvaWQ
uZ29vZ2xlYXBpcy5jb20vYXR0ZXN0YXRpb24vY3JsLzA2MzYwNzY4NGI4M2IxOTM2
Mzc1MDUwKBgKKwYBBAHWeQIBFQQaMBgCARkKAQACAQMKAQEECGFiY2RlZmdoMAoG
CCqGSM49BAMCA0gAMEUCIDadx4jX0NQtw77N8jnPQOqiTnkCONBQwN6Y1ryP9n0P
AiEAl5mB+GCvKN7EqtLfqBMEkuaFd4053AEi7+QZDCEvm7c=
-----END CERTIFICATE-----`,
];

function loadGoogleRoots(): crypto.X509Certificate[] {
  const roots: crypto.X509Certificate[] = [];
  for (const pem of GOOGLE_ROOT_CERTS_PEM) {
    try {
      roots.push(new crypto.X509Certificate(pem));
    } catch {
      // skip invalid roots
    }
  }
  return roots;
}

const GOOGLE_ROOTS = loadGoogleRoots();

function verifyChain(
  certs: crypto.X509Certificate[]
): [boolean, string, CertChainEntry[]] {
  const chainInfo = serializeCertChain(
    certs,
    "Google Hardware Attestation Root"
  );
  if (certs.length === 0) return [false, "empty certificate chain", chainInfo];

  // Check intermediate links
  for (const entry of chainInfo.slice(0, -1)) {
    if (!entry.signature_valid) {
      return [
        false,
        `chain verification failed at cert[${entry.index}]`,
        chainInfo,
      ];
    }
  }

  // Verify last cert against known Google root (cryptographic check only)
  const lastCert = certs[certs.length - 1];
  let rootVerified = false;
  for (const root of GOOGLE_ROOTS) {
    if (verifyCertSignature(lastCert, root.publicKey)) {
      rootVerified = true;
      break;
    }
  }

  if (chainInfo.length > 0) {
    chainInfo[chainInfo.length - 1].signature_valid = rootVerified;
  }

  if (!rootVerified) {
    return [
      false,
      "chain not rooted to a known Google hardware attestation root",
      chainInfo,
    ];
  }

  return [true, "", chainInfo];
}

export function verifyAndroidAttestation(
  certChainB64: string[],
  expectedChallenge: Buffer,
  userPublicKeyHex: string,
  requireBiometric = true
): [boolean, string, Record<string, unknown>] {
  const details: Record<string, unknown> = {};

  if (!certChainB64.length) {
    return [false, "no attestation chain provided", details];
  }

  // Parse certs
  let certs: crypto.X509Certificate[];
  try {
    certs = certChainB64.map((b64) => {
      const der = Buffer.from(b64, "base64");
      return new crypto.X509Certificate(der);
    });
    details.chain_length = certs.length;
  } catch (exc) {
    return [false, `invalid certificate in chain: ${exc}`, details];
  }

  // Verify chain
  const [chainOk, chainReason, chainInfo] = verifyChain(certs);
  details.chain_verified = chainOk;
  details.cert_chain = chainInfo;
  if (!chainOk) {
    details.failure_reason = chainReason;
    return [false, chainReason, details];
  }

  // Extract key attestation extension from leaf
  const leaf = certs[0];
  const extData = findExtensionValue(
    leaf.raw,
    ANDROID_KEY_ATTESTATION_OID
  );
  if (extData === null) {
    details.attestation_extension_present = false;
    return [
      false,
      "no key attestation extension; key is not hardware-attested",
      details,
    ];
  }
  details.attestation_extension_present = true;
  details.attestation_extension_len = extData.length;

  // Verify leaf certificate public key matches the registered public key
  try {
    const leafPubJwk = leaf.publicKey.export({ format: "jwk" }) as {
      x?: string;
      y?: string;
      kty?: string;
      crv?: string;
    };
    if (leafPubJwk.kty !== "EC" || !leafPubJwk.x || !leafPubJwk.y) {
      return [
        false,
        "leaf certificate does not contain an EC public key",
        details,
      ];
    }
    const leafX = Buffer.from(leafPubJwk.x, "base64url");
    const leafY = Buffer.from(leafPubJwk.y, "base64url");
    const leafRawHex = Buffer.concat([leafX, leafY]).toString("hex");
    details.leaf_public_key_hex = leafRawHex.substring(0, 20) + "...";
    if (leafRawHex !== userPublicKeyHex) {
      return [
        false,
        "leaf certificate public key does not match registered key",
        details,
      ];
    }
    details.public_key_binding = true;
  } catch (exc) {
    return [false, `failed to extract leaf public key: ${exc}`, details];
  }

  // Verify challenge is present in attestation extension
  const challengeHash = crypto
    .createHash("sha256")
    .update(expectedChallenge)
    .update(Buffer.from(userPublicKeyHex, "hex"))
    .digest();

  if (extData.includes(expectedChallenge)) {
    details.challenge_binding = "raw_challenge_found";
  } else if (extData.includes(challengeHash)) {
    details.challenge_binding = "challenge_hash_found";
  } else {
    return [
      false,
      "challenge not found in attestation extension",
      details,
    ];
  }

  // Parse KeyDescription for authorization details
  const keyDesc = parseKeyDescription(extData);
  if (!keyDesc.error) {
    const tee = keyDesc.teeEnforced;
    const sw = keyDesc.softwareEnforced;
    const userAuthTypeTee = typeof tee[TAG_USER_AUTH_TYPE] === "number" ? tee[TAG_USER_AUTH_TYPE] as number : 0;
    const userAuthTypeSw = typeof sw[TAG_USER_AUTH_TYPE] === "number" ? sw[TAG_USER_AUTH_TYPE] as number : 0;
    const noAuth = tee[TAG_NO_AUTH_REQUIRED] || sw[TAG_NO_AUTH_REQUIRED];
    const authTimeout = typeof tee[TAG_AUTH_TIMEOUT] === "number" ? tee[TAG_AUTH_TIMEOUT] as number : undefined;

    const biometricInTee = (userAuthTypeTee & AUTH_TYPE_FINGERPRINT) !== 0;
    const biometricInSw = (userAuthTypeSw & AUTH_TYPE_FINGERPRINT) !== 0;

    const secLevel = keyDesc.attestationSecurityLevel;
    details.key_protection = {
      security_level:
        secLevel === SECURITY_LEVEL_STRONGBOX ? "StrongBox" :
        secLevel === SECURITY_LEVEL_TEE ? "TEE" : "Software",
      biometric_required_tee: biometricInTee,
      biometric_required_sw: biometricInSw,
      no_auth_required: Boolean(noAuth),
      auth_timeout: authTimeout,
    };

    if (requireBiometric && !biometricInTee) {
      return [false, "key is not biometrically protected in TEE", details];
    }
  } else {
    details.key_description_parse_error = keyDesc.error;
    if (requireBiometric) {
      return [false, `could not verify biometric protection: ${keyDesc.error}`, details];
    }
  }

  details.valid = true;
  details.verified_at = Date.now() / 1000;
  return [true, "ok", details];
}
